File: crates/voidx-permission/src/engine.rs
//! Permission engine — combines sandbox + approval policy + rules + session whitelist.
//!
//! Ported from `src/voidx/permission/engine.py` + `service.py`.

use crate::error::PermissionError;
use crate::evaluate::PermissionVerdict;
use crate::sandbox::Sandbox;
use crate::schema::{self, Action, Ruleset};
use voidx_config::{ApprovalPolicy, SandboxMode};

/// Tool capability classification — mirrors Python's PermissionCapability.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ToolCapability {
    ReadTools,
    FileWrite,
    FileFormat,
    BashRead,
    BashWrite,
    AgentReadonly,
    AgentImplement,
    McpTools,
    Other,
}

impl ToolCapability {
    pub fn classify(tool_id: &str, tool_args: &serde_json::Value) -> Self {
        match tool_id {
            "read" | "glob" | "grep" | "webfetch" | "websearch" | "repo_map"
            | "lsp_diagnostics" | "lsp_symbols" | "lsp_definition" | "lsp_references"
            | "todo" | "task_status" => Self::ReadTools,
            "write" => Self::FileWrite,
            "edit" => Self::FileWrite,
            "lsp_format" => Self::FileFormat,
            "bash" => {
                // Check if the bash command is read-only
                if is_bash_readonly(tool_args) {
                    Self::BashRead
                } else {
                    Self::BashWrite
                }
            }
            "agent" => {
                let agent_type = tool_args
                    .get("agent_type")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                if agent_type == "implement" {
                    Self::AgentImplement
                } else {
                    Self::AgentReadonly
                }
            }
            id if id.starts_with("mcp__") || id.starts_with("mcp/") => Self::McpTools,
            _ => Self::Other,
        }
    }
}

/// Check if a bash command appears to be read-only.
fn is_bash_readonly(args: &serde_json::Value) -> bool {
    let command = args
        .get("command")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let lower = command.trim().to_lowercase();

    // Common read-only commands
    const READONLY_PREFIXES: &[&str] = &[
        "ls", "cat", "head", "tail", "find", "grep", "rg", "ag", "ack",
        "wc", "sort", "uniq", "diff", "which", "where", "type",
        "echo", "pwd", "whoami", "id", "uname", "date", "env",
        "git status", "git log", "git diff", "git show", "git branch",
        "cargo check", "cargo test", "cargo build", "cargo clippy",
        "npm test", "npm run", "npm list",
        "python -c", "python3 -c",
    ];

    for prefix in READONLY_PREFIXES {
        if lower.starts_with(prefix) {
            return true;
        }
    }

    // If it contains pipes or redirects to read, consider read-only
    if lower.contains("|") && !lower.contains(">") && !lower.contains(">>") {
        return true;
    }

    false
}

/// Classified tool call — mirrors Python's ClassifiedToolCall.
#[derive(Debug, Clone)]
pub struct ClassifiedToolCall {
    pub tool_call_id: String,
    pub name: String,
    pub args: serde_json::Value,
    pub capability: ToolCapability,
}

impl ClassifiedToolCall {
    pub fn classify(tool_call_id: &str, name: &str, args: &serde_json::Value) -> Self {
        Self {
            tool_call_id: tool_call_id.to_string(),
            name: name.to_string(),
            args: args.clone(),
            capability: ToolCapability::classify(name, args),
        }
    }
}

/// Permission decision — mirrors Python's PermissionDecision.
#[derive(Debug, Clone)]
pub struct PermissionDecision {
    pub action: Action,
    pub tool_call: ClassifiedToolCall,
    pub reason: String,
    pub failure_check: bool,
}

/// Permission context — mirrors Python's PermissionContext.
#[derive(Debug, Clone)]
pub struct PermissionContext {
    pub workspace: String,
    pub interaction_mode: String,
    pub permission_mode: String,
    pub sandbox_mode: SandboxMode,
    pub sandbox_workspace_write: Vec<String>,
    pub approval_policy: ApprovalPolicy,
    pub session_allow: Vec<String>,
    pub session_deny: Vec<String>,
}

/// The permission engine: sandbox + rules + session whitelist + approval policy.
#[derive(Debug, Clone)]
pub struct PermissionEngine {
    sandbox: Sandbox,
    approval: ApprovalPolicy,
    rules: Ruleset,
    session_allow: Vec<String>,
    session_deny: Vec<String>,
}

impl PermissionEngine {
    pub fn new(
        mode: SandboxMode,
        workspace_write: bool,
        approval: ApprovalPolicy,
        extra_paths: Vec<std::path::PathBuf>,
    ) -> Self {
        Self {
            sandbox: Sandbox {
                mode,
                workspace_write,
                extra_paths,
            },
            approval,
            rules: schema::basic_rules(),
            session_allow: Vec::new(),
            session_deny: Vec::new(),
        }
    }

    /// Pre-approve a tool for the entire session.
    pub fn allow(&mut self, tool: &str) {
        if !self.session_allow.contains(&tool.to_string()) {
            self.session_allow.push(tool.to_string());
        }
        self.session_deny.retain(|t| t != tool);
    }

    /// Pre-deny a tool for the entire session.
    pub fn deny(&mut self, tool: &str) {
        if !self.session_deny.contains(&tool.to_string()) {
            self.session_deny.push(tool.to_string());
        }
        self.session_allow.retain(|t| t != tool);
    }

    /// Check if a tool is in the session allow list.
    pub fn is_allowed(&self, tool: &str) -> bool {
        self.session_allow.contains(&tool.to_string())
    }

    /// Check if a tool is in the session deny list.
    pub fn is_denied(&self, tool: &str) -> bool {
        self.session_deny.contains(&tool.to_string())
    }

    /// Get the sandbox mode for reporting.
    pub fn sandbox_mode(&self) -> SandboxMode {
        self.sandbox.mode
    }

    /// Get the approval policy for reporting.
    pub fn approval_policy(&self) -> ApprovalPolicy {
        self.approval
    }

    /// Classify a tool call.
    pub fn classify(&self, tool_id: &str, args: &serde_json::Value) -> ClassifiedToolCall {
        ClassifiedToolCall::classify("", tool_id, args)
    }

    /// Full authorization flow — mirrors Python's authorize_tool_call.
    pub fn authorize(
        &self,
        tool_id: &str,
        args: &serde_json::Value,
        workspace: &std::path::Path,
        interaction_mode: &str,
        plan_mode: bool,
    ) -> PermissionDecision {
        let classified = ClassifiedToolCall::classify("", tool_id, args);

        // 1. Session deny list
        if self.session_deny.contains(&tool_id.to_string()) {
            return PermissionDecision {
                action: Action::Deny,
                tool_call: classified,
                reason: format!("Tool '{}' is denied for this session", tool_id),
                failure_check: false,
            };
        }

        // 2. Session allow list
        if self.session_allow.contains(&tool_id.to_string()) {
            return PermissionDecision {
                action: Action::Allow,
                tool_call: classified,
                reason: String::new(),
                failure_check: false,
            };
        }

        // 3. Sandbox check
        if let Err(e) = self.sandbox_check(tool_id, args, workspace) {
            return PermissionDecision {
                action: Action::Deny,
                tool_call: classified,
                reason: e.to_string(),
                failure_check: false,
            };
        }

        // 4. Plan mode check — block writes in plan mode
        if plan_mode || interaction_mode == "plan" {
            match classified.capability {
                ToolCapability::FileWrite
                | ToolCapability::FileFormat
                | ToolCapability::BashWrite
                | ToolCapability::AgentImplement => {
                    return PermissionDecision {
                        action: Action::Deny,
                        tool_call: classified,
                        reason: format!("Tool '{}' is blocked in plan mode", tool_id),
                        failure_check: false,
                    };
                }
                _ => {}
            }
        }

        // 5. Rule evaluation
        let tool_pattern = extract_tool_pattern(tool_id, args);
        let rule_action = schema::evaluate_rules(&self.rules, tool_id, &tool_pattern);

        // 6. Apply approval policy overlay
        let final_action = match self.approval {
            ApprovalPolicy::Never => Action::Allow,
            ApprovalPolicy::Untrusted => {
                // In untrusted mode, ask for dangerous tools
                match rule_action {
                    Action::Ask => {
                        match classified.capability {
                            ToolCapability::FileWrite | ToolCapability::BashWrite
                            | ToolCapability::AgentImplement | ToolCapability::McpTools => Action::Ask,
                            _ => Action::Allow,
                        }
                    }
                    other => other,
                }
            }
            ApprovalPolicy::OnFailure => {
                // Auto-allow non-bash tools; ask for bash
                match classified.capability {
                    ToolCapability::BashWrite => Action::Ask,
                    _ => rule_action,
                }
            }
            ApprovalPolicy::OnRequest => {
                // Auto-allow everything; agent explicitly requests if needed
                Action::Allow
            }
        };

        let failure_check = matches!(final_action, Action::Allow)
            && matches!(
                classified.capability,
                ToolCapability::FileWrite
                    | ToolCapability::BashWrite
                    | ToolCapability::AgentImplement
            );

        PermissionDecision {
            action: final_action,
            tool_call: classified,
            reason: String::new(),
            failure_check,
        }
    }

    /// Evaluate a tool call and return a verdict (backward compat).
    pub fn evaluate(
        &self,
        tool_id: &str,
        tool_args: &serde_json::Value,
        workspace: &std::path::Path,
    ) -> Result<PermissionVerdict, PermissionError> {
        let decision = self.authorize(tool_id, tool_args, workspace, "auto", false);
        match decision.action {
            Action::Allow => {
                if decision.failure_check {
                    Ok(PermissionVerdict::AllowWithFailureCheck {
                        reason: decision.reason,
                    })
                } else {
                    Ok(PermissionVerdict::Allow)
                }
            }
            Action::Ask => Ok(PermissionVerdict::AskUser(format!(
                "Tool '{}' requires approval",
                tool_id
            ))),
            Action::Deny => Ok(PermissionVerdict::Deny(
                decision.reason.or_default(format!("Tool '{}' denied", tool_id)),
            )),
        }
    }

    /// Sandbox check for a tool call.
    fn sandbox_check(
        &self,
        tool_id: &str,
        args: &serde_json::Value,
        workspace: &std::path::Path,
    ) -> Result<(), PermissionError> {
        match tool_id {
            "bash" => self.sandbox.check_bash(),
            "write" | "edit" => {
                if let Some(path) = args.get("file_path").and_then(|v| v.as_str()) {
                    self.sandbox
                        .check_write(&std::path::PathBuf::from(path), workspace)
                } else {
                    Ok(())
                }
            }
            _ => Ok(()),
        }
    }
}

/// Extract the tool pattern from args (e.g., agent_type for agent tool).
fn extract_tool_pattern(tool_id: &str, args: &serde_json::Value) -> String {
    if tool_id == "agent" {
        args.get("agent_type")
            .and_then(|v| v.as_str())
            .unwrap_or("*")
            .to_string()
    } else {
        "*".to_string()
    }
}

/// Helper trait for providing default reason.
trait OrDefault {
    fn or_default(self, default: String) -> String;
}

impl OrDefault for String {
    fn or_default(self, default: String) -> String {
        if self.is_empty() { default } else { self }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_read_tool_allowed_in_untrusted() {
        let engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"file_path": "test.rs"});

        let verdict = engine.evaluate("read", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::Allow));
    }

    #[test]
    fn test_write_tool_asks_in_untrusted() {
        let engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"file_path": "test.rs"});

        let verdict = engine.evaluate("write", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::AskUser(_)));
    }

    #[test]
    fn test_bash_asks_in_untrusted() {
        let engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"command": "rm -rf /"});

        let verdict = engine.evaluate("bash", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::AskUser(_)));
    }

    #[test]
    fn test_session_allow_overrides() {
        let mut engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        engine.allow("bash");
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"command": "ls"});

        let verdict = engine.evaluate("bash", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::Allow));
    }

    #[test]
    fn test_session_deny_overrides() {
        let mut engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        engine.deny("read");
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"file_path": "test.rs"});

        let verdict = engine.evaluate("read", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::Deny(_)));
    }

    #[test]
    fn test_plan_mode_blocks_writes() {
        let engine = PermissionEngine::new(
            SandboxMode::WorkspaceWrite,
            true,
            ApprovalPolicy::Untrusted,
            vec![],
        );
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"file_path": "test.rs"});

        let decision = engine.authorize("write", &args, ws.path(), "plan", true);
        assert_eq!(decision.action, Action::Deny);
    }

    #[test]
    fn test_never_policy_auto_allows() {
        let engine = PermissionEngine::new(
            SandboxMode::DangerFullAccess,
            true,
            ApprovalPolicy::Never,
            vec![],
        );
        let ws = tempdir().unwrap();
        let args = serde_json::json!({"command": "rm -rf /"});

        let verdict = engine.evaluate("bash", &args, ws.path()).unwrap();
        assert!(matches!(verdict, PermissionVerdict::Allow));
    }

    #[test]
    fn test_tool_capability_classification() {
        assert_eq!(
            ToolCapability::classify("read", &serde_json::json!({})),
            ToolCapability::ReadTools
        );
        assert_eq!(
            ToolCapability::classify("write", &serde_json::json!({})),
            ToolCapability::FileWrite
        );
        assert_eq!(
            ToolCapability::classify("agent", &serde_json::json!({"agent_type": "implement"})),
            ToolCapability::AgentImplement
        );
        assert_eq!(
            ToolCapability::classify("agent", &serde_json::json!({"agent_type": "explore"})),
            ToolCapability::AgentReadonly
        );
        assert_eq!(
            ToolCapability::classify("mcp__server__tool", &serde_json::json!({})),
            ToolCapability::McpTools
        );
    }
}

File: python/voidx/permission/engine.py
"""Central permission engine: capability classification, policies, and mode overlays."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum

from voidx.config import ApprovalPolicy, ApprovalReviewer, PermissionMode
from voidx.permission.evaluate import evaluate
from voidx.permission.sandbox import check_sandbox_bash, check_sandbox_filepath
from voidx.permission.schema import Action, Rule, Ruleset
from voidx.permission.wildcard import match as wildcard_match


class PermissionCapability(str, Enum):
    READ_TOOLS = "read_tools"
    FILE_WRITE = "file_write"
    FILE_FORMAT = "file_format"
    BASH_READ = "bash_read"
    BASH_WRITE = "bash_write"
    AGENT_READONLY = "agent_readonly"
    AGENT_IMPLEMENT = "agent_implement"
    MCP_TOOLS = "mcp_tools"
    OTHER = "other"


BASIC_RULES: Ruleset = [
    Rule(permission="read", pattern="*", action="allow"),
    Rule(permission="glob", pattern="*", action="allow"),
    Rule(permission="grep", pattern="*", action="allow"),
    Rule(permission="webfetch", pattern="*", action="allow"),
    Rule(permission="websearch", pattern="*", action="allow"),
    Rule(permission="todo", pattern="*", action="allow"),
    Rule(permission="task_status", pattern="*", action="allow"),
    Rule(permission="repo_map", pattern="*", action="allow"),
    Rule(permission="lsp_diagnostics", pattern="*", action="allow"),
    Rule(permission="lsp_symbols", pattern="*", action="allow"),
    Rule(permission="lsp_definition", pattern="*", action="allow"),
    Rule(permission="lsp_references", pattern="*", action="allow"),
    Rule(permission="agent", pattern="*", action="allow"),
    Rule(permission="write", pattern="*", action="ask"),
    Rule(permission="edit", pattern="*", action="ask"),
    Rule(permission="bash", pattern="*", action="ask"),
    Rule(permission="lsp_format", pattern="*", action="ask"),
    Rule(permission="agent", pattern="implement", action="ask"),
    Rule(permission="mcp__*", pattern="*", action="ask"),
    Rule(permission="mcp/*", pattern="*", action="ask"),
]


@dataclass(frozen=True)
class PermissionContext:
    workspace: str
    interaction_mode: str = "auto"
    permission_mode: str = PermissionMode.DEFAULT.value
    sandbox_mode: str = "workspace-write"
    sandbox_workspace_write: tuple[str, ...] = ()
    approval_policy: str = ApprovalPolicy.UNTRUSTED.value
    approval_reviewer: str = ApprovalReviewer.USER.value
    session_allow: frozenset[str] = field(default_factory=frozenset)
    session_deny: frozenset[str] = field(default_factory=frozenset)

    @classmethod
    def from_service(
        cls,
        service,
        *,
        workspace: str,
        interaction_mode: str | None = None,
        plan_mode: bool = False,
    ) -> "PermissionContext":
        mode = interaction_mode or "auto"
        if plan_mode:
            mode = "plan"
        return cls(
            workspace=workspace,
            interaction_mode=mode,
            permission_mode=getattr(service, "permission_mode", PermissionMode.DEFAULT.value),
            sandbox_mode=getattr(service, "sandbox_mode", "workspace-write"),
            sandbox_workspace_write=tuple(getattr(service, "sandbox_workspace_write", []) or []),
            approval_policy=getattr(service, "approval_policy", ApprovalPolicy.UNTRUSTED.value),
            approval_reviewer=getattr(service, "approval_reviewer", ApprovalReviewer.USER.value),
            session_allow=frozenset(getattr(service, "_session_allow", set())),
            session_deny=frozenset(getattr(service, "_session_deny", set())),
        )


@dataclass(frozen=True)
class ClassifiedToolCall:
    tool_call: dict
    name: str
    args: dict
    pattern: str
    capability: PermissionCapability


@dataclass(frozen=True)
class PermissionDecision:
    action: Action
    tool_call: dict
    name: str
    args: dict
    pattern: str
    capability: PermissionCapability
    source: str
    reason: str = ""
    failure_check: bool = False


def authorize_tool_call(tool_call: dict, context: PermissionContext) -> PermissionDecision:
    classified = classify_tool_call(tool_call)

    reason = sandbox_denial_reason(classified, context)
    if reason:
        return _decision(classified, "deny", "sandbox", reason)

    reason = mode_overlay_denial_reason(classified, context)
    if reason:
        return _decision(classified, "deny", "mode", reason)

    session_action = session_action_for_tool(classified.name, context)
    if session_action:
        reason = _reason_for(classified, session_action)
        return _decision(classified, session_action, "session", reason)

    action = strategy_action_for_tool(classified, context)
    if action != "ask":
        return _decision(classified, action, "strategy", _reason_for(classified, action))

    return resolve_approval(classified, context)


def decide_base_action(tool: str, pattern: str, context: PermissionContext) -> Action:
    classified = classify_tool_call(tool_call_from_pattern(tool, pattern))
    session_action = session_action_for_tool(classified.name, context)
    if session_action:
        return session_action
    return strategy_action_for_tool(classified, context)


def classify_tool_call(tool_call: dict) -> ClassifiedToolCall:
    name = repair_tool_name(str(tool_call.get("name", "")))
    args = tool_call.get("args", {})
    if not isinstance(args, dict):
        args = {}
    repaired = {**tool_call, "name": name, "args": args}
    pattern = build_pattern(name, args)
    return ClassifiedToolCall(
        tool_call=repaired,
        name=name,
        args=args,
        pattern=pattern,
        capability=_capability(name, args),
    )


def tool_call_from_pattern(tool: str, pattern: str = "*") -> dict:
    name = repair_tool_name(tool)
    if name == "bash":
        args = {"command": pattern}
    elif name == "agent":
        args = {"agent": pattern}
    elif name in _FILE_PATTERN_TOOLS:
        args = {"file_path": pattern}
    else:
        args = {}
    return {"name": name, "args": args}


def sandbox_denial_reason(classified: ClassifiedToolCall, context: PermissionContext) -> str | None:
    if context.sandbox_mode == "danger-full-access":
        return None

    if context.sandbox_mode == "read-only":
        if classified.capability in {
            PermissionCapability.FILE_WRITE,
            PermissionCapability.FILE_FORMAT,
            PermissionCapability.BASH_WRITE,
        }:
            return f"SANDBOX READ-ONLY: '{classified.name}' is not allowed."
        if classified.capability == PermissionCapability.AGENT_IMPLEMENT:
            return "SANDBOX READ-ONLY: cannot delegate to implement."
        return None

    if context.sandbox_mode == "workspace-write":
        if classified.capability in {PermissionCapability.FILE_WRITE, PermissionCapability.FILE_FORMAT}:
            file_path = classified.args.get("file_path", "")
            if file_path:
                return check_sandbox_filepath(
                    file_path,
                    context.workspace,
                    list(context.sandbox_workspace_write),
                )
        if classified.name == "bash":
            command = classified.args.get("command", "")
            if command:
                return check_sandbox_bash(
                    command,
                    context.workspace,
                    list(context.sandbox_workspace_write),
                )
    return None


def mode_overlay_denial_reason(classified: ClassifiedToolCall, context: PermissionContext) -> str | None:
    if context.interaction_mode != "plan":
        return None
    if classified.capability in {
        PermissionCapability.FILE_WRITE,
        PermissionCapability.FILE_FORMAT,
        PermissionCapability.BASH_WRITE,
    }:
        return f"BLOCKED by plan mode: '{classified.name}' is not allowed."
    if classified.capability == PermissionCapability.AGENT_IMPLEMENT:
        return "BLOCKED by plan mode: cannot delegate to implement."
    return None


def session_action_for_tool(tool: str, context: PermissionContext) -> Action | None:
    if any(_session_rule_matches(tool, rule) for rule in context.session_deny):
        return "deny"
    if any(_session_rule_matches(tool, rule) for rule in context.session_allow):
        return "allow"
    return None


def strategy_action_for_tool(classified: ClassifiedToolCall, context: PermissionContext) -> Action:
    if context.permission_mode == PermissionMode.ACCEPT_EDITS.value and classified.capability in {
        PermissionCapability.FILE_WRITE,
        PermissionCapability.FILE_FORMAT,
    }:
        return "allow"
    if classified.capability == PermissionCapability.BASH_READ:
        return "allow"
    return evaluate(classified.name, classified.pattern, BASIC_RULES).action


def resolve_approval(classified: ClassifiedToolCall, context: PermissionContext) -> PermissionDecision:
    policy = context.approval_policy
    if policy in {ApprovalPolicy.NEVER.value, ApprovalPolicy.ON_REQUEST.value}:
        return _decision(classified, "allow", "approval_policy", _reason_for(classified, "allow"))

    if policy == ApprovalPolicy.ON_FAILURE.value:
        if classified.capability == PermissionCapability.BASH_WRITE:
            return _decision(classified, "ask", "approval_policy", _reason_for(classified, "ask"))
        return _decision(
            classified,
            "allow",
            "approval_policy",
            _reason_for(classified, "allow"),
            failure_check=True,
        )

    if context.approval_reviewer == ApprovalReviewer.AUTO_REVIEW.value:
        if classified.capability == PermissionCapability.BASH_WRITE:
            return _decision(classified, "ask", "auto_review", _reason_for(classified, "ask"))
        return _decision(classified, "allow", "auto_review", _reason_for(classified, "allow"))

    return _decision(classified, "ask", "strategy", _reason_for(classified, "ask"))


def repair_tool_name(tool: str) -> str:
    tool_map = {
        "Read": "read", "Write": "write", "Edit": "edit",
        "MultiEdit": "edit", "multiEdit": "edit", "multi_edit": "edit",
        "Glob": "glob", "Grep": "grep", "Bash": "bash",
        "Agent": "agent", "TodoWrite": "todo", "Todo": "todo",
        "WebFetch": "webfetch", "WebSearch": "websearch",
        "read_file": "read", "write_file": "write",
        "edit_file": "edit", "shell": "bash",
        "readfile": "read", "writefile": "write",
        "search": "grep", "find": "glob",
        "RepoMap": "repo_map", "repomap": "repo_map", "Repo_map": "repo_map",
        "LspDiagnostics": "lsp_diagnostics", "LspSymbols": "lsp_symbols",
        "LspDefinition": "lsp_definition", "LspReferences": "lsp_references",
        "LspFormat": "lsp_format",
    }
    return tool_map.get(tool, tool_map.get(tool.lower(), tool))


def build_pattern(tool: str, args: dict) -> str:
    if tool == "bash":
        return str(args.get("command", "*"))
    if tool in _FILE_PATTERN_TOOLS:
        return str(args.get("file_path", "*"))
    if tool == "agent":
        return delegated_agent(args) or "*"
    return "*"


def delegated_agent(args: dict) -> str:
    return str(args.get("agent") or "")


def is_safe_bash(command: str) -> bool:
    stripped = command.strip()
    if not stripped or stripped.startswith("#"):
        return True
    if re.search(r" > ", stripped) or re.search(r" >> ", stripped):
        return False
    if re.search(r"\|\s*tee\b", stripped):
        return False

    words = stripped.split()
    prog = words[0].lower()

    if prog == "git" and len(words) > 1:
        sub = words[1]
        read_only_git = {
            "status", "log", "diff", "show", "blame", "rev-parse", "rev-list",
            "ls-files", "ls-tree", "describe", "shortlog", "reflog", "cherry",
            "whatchanged", "notes", "grep", "bisect",
            "config", "stash", "branch", "tag", "remote", "worktree",
        }
        if sub not in read_only_git:
            return False
        if sub == "stash":
            return len(words) > 2 and words[2] in ("list", "show")
        if sub == "bisect":
            return len(words) > 2 and words[2] in ("log", "view", "visualize")
        if sub in ("branch", "tag"):
            return "-d" not in words and "-D" not in words
        if sub == "remote":
            return "-v" in words or "--verbose" in words or len(words) == 2
        if sub == "worktree":
            return len(words) > 2 and words[2] == "list"
        return True

    if prog == "gh" and len(words) > 1:
        sub = words[1]
        if sub == "pr":
            return len(words) > 2 and words[2] in ("view", "list", "status", "checks", "diff")
        if sub == "issue":
            return len(words) > 2 and words[2] in ("view", "list", "status")
        if sub == "api":
            cmd_upper = stripped.upper()
            if "-X" in cmd_upper or "--METHOD" in cmd_upper:
                return "GET" in cmd_upper
            return True
        if sub in ("auth", "config", "completion", "secret"):
            return len(words) == 2 or (len(words) > 2 and words[2] in ("list", "status", "view"))
        return False

    read_only = {
        "ls", "dir", "cat", "head", "tail", "wc", "which", "where", "whereis",
        "echo", "printf", "pwd", "date", "whoami", "uname", "env", "printenv",
        "df", "du", "sort", "uniq", "cut", "tr", "column", "less", "more",
        "find", "grep", "egrep", "fgrep", "rg", "file", "stat", "od",
        "true", "false", "test", "[", "type", "basename", "dirname",
        "realpath", "readlink", "hostname", "id", "groups", "logname",
        "uptime", "free", "swapon", "lscpu", "lsblk", "lspci", "lsusb",
    }
    if prog in read_only:
        return True

    if prog in ("pip", "pip3") and len(words) > 1:
        return words[1] in ("list", "show", "freeze", "config", "cache")
    if prog in ("npm", "npx") and len(words) > 1:
        return words[1] in ("list", "ls", "view", "info", "outdated")
    if prog == "cargo" and len(words) > 1:
        return words[1] in ("search", "doc", "readme")
    if prog == "go" and len(words) > 1:
        return words[1] in ("list", "doc", "version", "env")

    return False


def _capability(tool: str, args: dict) -> PermissionCapability:
    if tool in {
        "read", "glob", "grep", "webfetch", "websearch", "todo", "task_status",
        "repo_map", "lsp_diagnostics", "lsp_symbols", "lsp_definition",
        "lsp_references",
    }:
        return PermissionCapability.READ_TOOLS
    if tool in {"write", "edit"}:
        return PermissionCapability.FILE_WRITE
    if tool == "lsp_format":
        return PermissionCapability.FILE_FORMAT
    if tool == "bash":
        return PermissionCapability.BASH_READ if is_safe_bash(str(args.get("command", ""))) else PermissionCapability.BASH_WRITE
    if tool == "agent":
        return PermissionCapability.AGENT_IMPLEMENT if delegated_agent(args) == "implement" else PermissionCapability.AGENT_READONLY
    if tool.startswith("mcp__") or tool.startswith("mcp/"):
        return PermissionCapability.MCP_TOOLS
    return PermissionCapability.OTHER


def _session_rule_matches(tool: str, rule: str) -> bool:
    if wildcard_match(tool, rule):
        return True
    if rule.startswith("mcp/"):
        return wildcard_match(tool, rule.replace("/", "__"))
    return False


def _decision(
    classified: ClassifiedToolCall,
    action: Action,
    source: str,
    reason: str = "",
    *,
    failure_check: bool = False,
) -> PermissionDecision:
    return PermissionDecision(
        action=action,
        tool_call=classified.tool_call,
        name=classified.name,
        args=classified.args,
        pattern=classified.pattern,
        capability=classified.capability,
        source=source,
        reason=reason,
        failure_check=failure_check,
    )


def _reason_for(classified: ClassifiedToolCall, action: Action) -> str:
    if action == "deny":
        return f"Permission denied: {classified.name} → {classified.pattern}"
    if action == "allow":
        return f"Permission allowed: {classified.name} → {classified.pattern}"
    return f"Permission required: {classified.name} → {classified.pattern}"


_FILE_PATTERN_TOOLS = {
    "read", "write", "edit",
    "lsp_diagnostics", "lsp_symbols", "lsp_definition",
    "lsp_references", "lsp_format",
}
